src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java

```diff
@@ -163,16 +184,91 @@ public class JGitAPIImpl extends LegacyCompatibleGitAPIImpl {
          * CliGitAPIImpl behavior */
         super(workspace == null ? new File(".") : workspace, hostKeyFactory);
         this.listener = listener;
+        hostKeyVerifierFactory.set(hostKeyFactory);
+        listenerThreadLocal.set(listener);
+    }
 
-        // to avoid rogue plugins from clobbering what we use, always
-        // make a point of overwriting it with ours.
-        SshSessionFactory.setInstance(new TrileadSessionFactory(hostKeyFactory, listener));
-
-        if (httpConnectionFactory != null) {
-            httpConnectionFactory.setCredentialsProvider(asSmartCredentialsProvider());
-            // allow override of HttpConnectionFactory to avoid JENKINS-37934
-            HttpTransport.setConnectionFactory(httpConnectionFactory);
+    private static SshdSessionFactory buildSshdSessionFactory() {
+        if (Files.notExists(SshHostKeyVerificationStrategy.JGIT_KNOWN_HOSTS_FILE.toPath())) {
+            try {
+                Files.createDirectories(SshHostKeyVerificationStrategy.JGIT_KNOWN_HOSTS_FILE
+                        .getParentFile()
+                        .toPath());
+                Files.createFile(SshHostKeyVerificationStrategy.JGIT_KNOWN_HOSTS_FILE.toPath());
+            } catch (IOException e) {
+                LOGGER.log(Level.SEVERE, "cannot create knowhosts file", e);
+            }
         }
+        return new SshdSessionFactory() {
+            private Path tmpKey;
+
+            @Override
+            protected SshConfigStore createSshConfigStore(File homeDir, File configFile, String localUserName) {
+                return new OpenSshConfigFile(homeDir, configFile, localUserName) {
+                    @Override
+                    public HostEntry lookup(String hostName, int port, String userName) {
+                        HostEntry hostEntry = super.lookup(hostName, port, userName);
+                        return hostKeyVerifierFactory.get().forJGit(null).customizeHostEntry(hostEntry);
+                    }
+                };
+            }
+
+            @Override
+            public File getHomeDirectory() {
+                return SystemUtils.getUserHome();
+            }
+
+            @Override
+            public File getSshDirectory() {
+                return SshHostKeyVerificationStrategy.JGIT_KNOWN_HOSTS_FILE.getParentFile();
+            }
+
+            @Override
+            protected Iterable<KeyPair> getDefaultKeys(File sshDir) {
+                SmartCredentialsProvider smartCredentialsProvider = JGitAPIImpl.getProvider();
+                // credentials have been added for the target repo only so we assume we have only the good one here
+                SSHUserPrivateKey sshUserPrivateKey = smartCredentialsProvider.getCredentials().values().stream()
```

Comment:
ideally, here we should have access to the target URL to use only this credential ()